src/main/scala/cromwell/engine/workflow/CallMetadataBuilder.scala

```diff
@@ -0,0 +1,190 @@
+package cromwell.engine.workflow
+
+import cromwell.binding._
+import cromwell.engine.backend.{CallMetadata, StdoutStderr}
+import cromwell.engine.db.ExecutionDatabaseKey
+import cromwell.engine.db.slick._
+import cromwell.engine.ExecutionIndex._
+
+import scala.language.postfixOps
+
+/**
+ * Builds call metadata suitable for return as part of a workflow metadata request.
+ */
+object CallMetadataBuilder {
+
+  // Collects up all the data for a single CallMetadata in the easiest-to-assemble format, but
+  // not quite the format required for the metadata REST endpoint.
+  private case class AssembledCallMetadata(key: ExecutionDatabaseKey,
+                                           execution: Execution,
+                                           inputs: Traversable[Symbol] = Seq.empty,
+                                           outputs: Option[Traversable[Symbol]] = None,
+                                           streams: Option[StdoutStderr] = None,
+                                           backend: Option[String] = None,
+                                           jobId: Option[String] = None,
+                                           backendStatus: Option[String] = None)
+
+  // Types used in interim steps of the construction of the call metadata.
+  // Map from an `ExecutionDatabaseKey` to the interim `AssembledCallMetadata` format.
+  private type ExecutionMap = Map[ExecutionDatabaseKey, AssembledCallMetadata]
```

Comment:
The first type alias is a concise expression of a `Map` with two fixed type variables, the second is a function from an argument of the first type alias to a result of the same type.  I think these add value in abstracting out the concepts and allowing the code to be much more concise, and as you've observed they aren't visible outside this class.  Also I don't see nice ways of making these simple case classes.

But I'm the author and not a reviewer, so I'm certainly biased... :smile: